File: 0001-two-sum/solution.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; ++i) {
          if (map.containsKey(target - nums[i])) {
            int[] line = new int[]{map.get(target - nums[i]), i};
            return line;
          }
          map.put(nums[i], i);
        }
        return new int[]{};
    }
}


class Main {

    static java.util.List<String> splitCases(String content) {
        content = content.replace("\r\n", "\n").replace("\r", "\n");
        String[] parts = content.split("\n\n");
        java.util.List<String> out = new java.util.ArrayList<>();
        for (String p : parts) {
            p = p.trim();
            if (!p.isEmpty()) out.add(p);
        }
        return out;
    }

    static class CaseData {
        java.util.List<String> inLines;
        String outStr;
        CaseData(java.util.List<String> inLines, String outStr) {
            this.inLines = inLines;
            this.outStr = outStr;
        }
    }

    static CaseData readCase(String block) {
        String[] lines = block.split("\n");
        String mode = "";
        java.util.List<String> ins = new java.util.ArrayList<>();
        StringBuilder outs = new StringBuilder();
        for (String ln : lines) {
            String t = ln.trim();
            if (t.equals("input:")) { mode = "in"; continue; }
            if (t.equals("output:")) { mode = "out"; continue; }
            if (mode.equals("in")) {
                ins.add(t);
            } else if (mode.equals("out")) {
                if (!t.isEmpty()) outs.append(t);
            }
        }
        return new CaseData(ins, outs.toString().trim());
    }

    static String trimBrackets(String s) {
        s = s.trim();
        if (s.startsWith("[") && s.endsWith("]") && s.length() >= 2) {
            return s.substring(1, s.length() - 1).trim();
        }
        return s;
    }

    static java.util.List<String> splitTopLevel(String inner) {
        java.util.List<String> items = new java.util.ArrayList<>();
        StringBuilder buf = new StringBuilder();
        int depth = 0;
        boolean inStr = false;
        boolean esc = false;

        for (int i = 0; i < inner.length(); i++) {
            char ch = inner.charAt(i);

            if (inStr) {
                buf.append(ch);
                if (esc) esc = false;
                else if (ch == '\\') esc = true;
                else if (ch == '"') inStr = false;
                continue;
            }

            if (ch == '"') {
                inStr = true;
                buf.append(ch);
                continue;
            }
            if (ch == '[') { depth++; buf.append(ch); continue; }
            if (ch == ']') { depth--; buf.append(ch); continue; }

            if (ch == ',' && depth == 0) {
                String t = buf.toString().trim();
                if (!t.isEmpty()) items.add(t);
                buf.setLength(0);
                continue;
            }
            buf.append(ch);
        }
        String tail = buf.toString().trim();
        if (!tail.isEmpty()) items.add(tail);
        return items;
    }

    static String parseString(String s) {
        s = s.trim();
        if (s.length() >= 2) {
            char a = s.charAt(0), b = s.charAt(s.length()-1);
            if ((a=='"' && b=='"') || (a=='\'' && b=='\'')) {
                return s.substring(1, s.length()-1);
            }
        }
        return s;
    }

    static boolean parseBool(String s) {
        s = s.trim().toLowerCase();
        return s.equals("true") || s.equals("1");
    }

    static int parseInt(String s) { return Integer.parseInt(s.trim()); }
    static long parseLong(String s) { return Long.parseLong(s.trim()); }
    static double parseDouble(String s) { return Double.parseDouble(s.trim()); }

    static int[] parseIntArray(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new int[0];
        java.util.List<String> parts = splitTopLevel(inner);
        int[] out = new int[parts.size()];
        for (int i=0;i<parts.size();i++) out[i] = parseInt(parts.get(i));
        return out;
    }

    static int[][] parseInt2D(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new int[0][];
        java.util.List<String> parts = splitTopLevel(inner);
        int[][] out = new int[parts.size()][];
        for (int i=0;i<parts.size();i++) out[i] = parseIntArray(parts.get(i));
        return out;
    }

    static long[] parseLongArray(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new long[0];
        java.util.List<String> parts = splitTopLevel(inner);
        long[] out = new long[parts.size()];
        for (int i=0;i<parts.size();i++) out[i] = parseLong(parts.get(i));
        return out;
    }

    static long[][] parseLong2D(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new long[0][];
        java.util.List<String> parts = splitTopLevel(inner);
        long[][] out = new long[parts.size()][];
        for (int i=0;i<parts.size();i++) out[i] = parseLongArray(parts.get(i));
        return out;
    }

    static double[] parseDoubleArray(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new double[0];
        java.util.List<String> parts = splitTopLevel(inner);
        double[] out = new double[parts.size()];
        for (int i=0;i<parts.size();i++) out[i] = parseDouble(parts.get(i));
        return out;
    }

    static double[][] parseDouble2D(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new double[0][];
        java.util.List<String> parts = splitTopLevel(inner);
        double[][] out = new double[parts.size()][];
        for (int i=0;i<parts.size();i++) out[i] = parseDoubleArray(parts.get(i));
        return out;
    }

    static boolean[] parseBoolArray(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new boolean[0];
        java.util.List<String> parts = splitTopLevel(inner);
        boolean[] out = new boolean[parts.size()];
        for (int i=0;i<parts.size();i++) out[i] = parseBool(parts.get(i));
        return out;
    }

    static boolean[][] parseBool2D(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new boolean[0][];
        java.util.List<String> parts = splitTopLevel(inner);
        boolean[][] out = new boolean[parts.size()][];
        for (int i=0;i<parts.size();i++) out[i] = parseBoolArray(parts.get(i));
        return out;
    }

    static String[] parseStringArray(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new String[0];
        java.util.List<String> parts = splitTopLevel(inner);
        String[] out = new String[parts.size()];
        for (int i=0;i<parts.size();i++) out[i] = parseString(parts.get(i));
        return out;
    }

    static String[][] parseString2D(String s) {
        String inner = trimBrackets(s);
        if (inner.isEmpty()) return new String[0][];
        java.util.List<String> parts = splitTopLevel(inner);
        String[][] out = new String[parts.size()][];
        for (int i=0;i<parts.size();i++) out[i] = parseStringArray(parts.get(i));
        return out;
    }

    static boolean equalDoubleArray(double[] a, double[] b) {
        if (a == null || b == null) return a == b;
        if (a.length != b.length) return false;
        for (int i=0;i<a.length;i++) if (Math.abs(a[i]-b[i]) > 1e-6) return false;
        return true;
    }

    static boolean equalDouble2D(double[][] a, double[][] b) {
        if (a == null || b == null) return a == b;
        if (a.length != b.length) return false;
        for (int i=0;i<a.length;i++) if (!equalDoubleArray(a[i], b[i])) return false;
        return true;
    }

    public static void main(String[] args) throws Exception {
        java.nio.file.Path p = java.nio.file.Paths.get("testcases.txt");
        if (!java.nio.file.Files.exists(p)) {
            System.out.println("testcases.txt not found");
            return;
        }
        byte[] bytes = java.nio.file.Files.readAllBytes(p);
        String content = new String(bytes, java.nio.charset.StandardCharsets.UTF_8);

        java.util.List<String> cases = splitCases(content);

        for (int ci=0; ci<cases.size(); ci++) {
            CaseData c = readCase(cases.get(ci));
            java.util.List<String> inLines = c.inLines;
            String outStr = c.outStr;

            if (inLines.size() < 2) {
                System.out.printf("Test %d: 输入行数不足，期望 2 行，实际 %d 行\n", ci+1, inLines.size());
                continue;
            }

            int[] arg0 = parseIntArray(inLines.get(0));
            int arg1 = parseInt(inLines.get(1));
            var got = new Solution().twoSum(arg0, arg1);

            if (!outStr.isEmpty()) {
            int[] expected = parseIntArray(outStr);
            boolean passed = java.util.Arrays.equals(got, expected);
                System.out.println("Test " + (ci+1) + ": " + (passed ? "Passed" : "Failed"));
                System.out.println(" Input(raw): " + inLines.subList(0, 2));
                System.out.println(" Expected(raw): " + outStr);
                System.out.println();
            } else {
                System.out.println("Test " + (ci+1) + ": Done (no expected output)");
                System.out.println();
            }
        }
    }
}

File: 0001-two-sum/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <cmath>
#include <regex>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
          for (int j = i + 1; j < n; ++j) {
            if (nums[i] + nums[j] == target) {
              return {i, j};
            }
          }
        }
        return {};
    }
};


static inline string trim(const string& s) {
    size_t i = 0, j = s.size();
    while (i < j && isspace((unsigned char)s[i])) i++;
    while (j > i && isspace((unsigned char)s[j-1])) j--;
    return s.substr(i, j - i);
}

static inline string trimBrackets(string s) {
    s = trim(s);
    if (s.size() >= 2 && s.front() == '[' && s.back() == ']') {
        return trim(s.substr(1, s.size()-2));
    }
    return s;
}

static vector<string> splitTopLevel(const string& inner) {
    vector<string> items;
    string buf;
    int depth = 0;
    bool inStr = false, esc = false;
    for (char ch : inner) {
        if (inStr) {
            buf.push_back(ch);
            if (esc) esc = false;
            else if (ch == '\\') esc = true;
            else if (ch == '"') inStr = false;
            continue;
        }
        if (ch == '"') { inStr = true; buf.push_back(ch); continue; }
        if (ch == '[') { depth++; buf.push_back(ch); continue; }
        if (ch == ']') { depth--; buf.push_back(ch); continue; }
        if (ch == ',' && depth == 0) {
            items.push_back(trim(buf));
            buf.clear();
            continue;
        }
        buf.push_back(ch);
    }
    if (!trim(buf).empty()) items.push_back(trim(buf));
    return items;
}

static string parseString(string s) {
    s = trim(s);
    if (s.size() >= 2) {
        char a = s.front(), b = s.back();
        if ((a=='"' && b=='"') || (a=='\'' && b=='\'')) {
            return s.substr(1, s.size()-2);
        }
    }
    return s;
}
static bool parseBool(string s) {
    s = trim(s);
    for (auto& c: s) c = (char)tolower(c);
    return s == "true" || s == "1";
}
static int parseInt(string s) { return stoi(trim(s)); }
static long long parseLong(string s) { return stoll(trim(s)); }
static double parseDouble(string s) { return stod(trim(s)); }

static vector<int> parseIntVec(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<int> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseInt(p));
    return out;
}
static vector<vector<int>> parseInt2D(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<vector<int>> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseIntVec(p));
    return out;
}

static vector<long long> parseLongVec(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<long long> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseLong(p));
    return out;
}
static vector<vector<long long>> parseLong2D(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<vector<long long>> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseLongVec(p));
    return out;
}

static vector<double> parseDoubleVec(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<double> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseDouble(p));
    return out;
}
static vector<vector<double>> parseDouble2D(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<vector<double>> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseDoubleVec(p));
    return out;
}

static vector<bool> parseBoolVec(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<bool> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseBool(p));
    return out;
}
static vector<vector<bool>> parseBool2D(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<vector<bool>> out;
    out.reserve(parts.size());
    for (auto &p : parts) {
        auto v = parseBoolVec(p);
        out.emplace_back(v.begin(), v.end());
    }
    return out;
}

static vector<string> parseStringVec(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<string> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseString(p));
    return out;
}
static vector<vector<string>> parseString2D(string s) {
    string inner = trimBrackets(s);
    if (inner.empty()) return {};
    auto parts = splitTopLevel(inner);
    vector<vector<string>> out;
    out.reserve(parts.size());
    for (auto &p : parts) out.push_back(parseStringVec(p));
    return out;
}

static bool equalDoubleVec(const vector<double>& a, const vector<double>& b) {
    if (a.size() != b.size()) return false;
    for (size_t i=0;i<a.size();i++) if (fabs(a[i]-b[i]) > 1e-6) return false;
    return true;
}
static bool equalDouble2D(const vector<vector<double>>& a, const vector<vector<double>>& b) {
    if (a.size() != b.size()) return false;
    for (size_t i=0;i<a.size();i++) if (!equalDoubleVec(a[i], b[i])) return false;
    return true;
}

static vector<string> splitCases(const string& content) {
    string s = content;
    // normalize newlines
    s.erase(std::remove(s.begin(), s.end(), '\r'), s.end());

    vector<string> cases;
    size_t start = 0;
    while (true) {
        size_t pos = s.find("\n\n", start);
        string blk = (pos == string::npos) ? s.substr(start) : s.substr(start, pos-start);
        blk = trim(blk);
        if (!blk.empty()) cases.push_back(blk);
        if (pos == string::npos) break;
        start = pos + 2;
    }
    return cases;
}

static pair<vector<string>, string> readCase(const string& block) {
    vector<string> inLines;
    string outStr;
    string mode;
    stringstream ss(block);
    string line;
    while (getline(ss, line)) {
        string t = trim(line);
        if (t == "input:") { mode = "in"; continue; }
        if (t == "output:") { mode = "out"; continue; }
        if (mode == "in") inLines.push_back(t);
        else if (mode == "out") if (!t.empty()) outStr += t;
    }
    return {inLines, trim(outStr)};
}

int main() {
    ifstream f("testcases.txt");
    if (!f.is_open()) {
        cout << "testcases.txt not found\n";
        return 0;
    }
    string content((istreambuf_iterator<char>(f)), istreambuf_iterator<char>());
    auto cases = splitCases(content);

    for (int ci=0; ci<(int)cases.size(); ci++) {
        auto [inLines, outStr] = readCase(cases[ci]);
        if ((int)inLines.size() < 2) {
            cout << "Test " << (ci+1) << ": 输入行数不足，期望 2 行，实际 " << inLines.size() << " 行\n";
            continue;
        }

        std::vector<int> arg0 = parseIntVec(inLines[0]);
        int arg1 = parseInt(inLines[1]);
        auto got = Solution().twoSum(arg0, arg1);
        std::vector<int> expected = parseIntVec(outStr);
        bool passed = (got == expected);

        cout << "Test " << (ci+1) << ": " << (passed ? "Passed" : "Failed") << "\n";
        cout << " Input(raw): ";
        for (int k = 0; k < inLines.size(); k++) cout << inLines[k] << (k+1==inLines.size()? "" : " | ");
        cout << "\n";
        cout << " Expected(raw): " << outStr << "\n";
        cout << " Done\n\n";
    }
    return 0;
}
